docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+contract. Can be called by anyone.
+
+==== `topUpKeep(address operator) external`
+
+Propagates information about stake top-up from the legacy KEEP staking contract
+to T staking contract. Can be called by anyone.
+
+==== `topUpNu(address operator) external`
+
+Propagates information about stake top-up from the legacy NU staking contract
+to T staking contract. Can be called by anyone.
+
+=== Undelegating a stake
+
+==== `cancelStake(address operator) external`
+
+Cancels T stake delegation. Can not be called in the same transaction as the
+original delegation order. Reverts after 2 hours from the time the delegation
+order was created. Clears up information about the operator so that the
+operator's address can be reused. Can be called by owner and operator.
+
+==== `undelegate(address operator) external`
+
+Orders undelegation from the given operator. Sets the delegation state as
+undelegating. Can be called only by the owner or operator.
+
+=== Recovering a stake
+
+==== `recoverStake(address operator) external`
+
+Sets the undelegating delegation state as recovered once the undelegation period
+passes. For liquid T delegations, transfers tokens back to the original owner.
+Can be called by anyone.
+
+=== Keeping information in sync
+
+==== `notifyKeepStakeDiscrepancy(address operator)`
+
+Notifies about the discrepancy between legacy KEEP stake and stake amount in T
+staking contract. Can be called by anyone, notifier receives a reward.
+
+Optionally: reward withdrawal can be split into a separate function to protect
+against MEV frontrunners. 
+
+==== `notifyNuStakeDiscrepancy(address operator)`
+
+Notifies about the discrepancy between legacy NU stake and stake amount in T
+staking contract. Can be called by anyone, notifier receives a reward.
+
+Optionally: reward withdrawal can be split into a separate function to protect
+against MEV frontrunners. 
+
+==== `setStakeDiscrepancyPenalty(uint256 penalty, unit256 rewardMultiplier) external onlyGovernance`
+
+Sets the penalty amount for stake discrepancy and reward multiplier for
+reporting it. The penalty is seized from the operator account, and 5% of the
+penalty, scaled by the multiplier, is given to the notifier. The rest of the
+tokens are burned. Can only be called by the governance. See `seize` function.
+
+=== Slashing a stake
+
+==== `slash(uint256 amount, address[] memory operators) external onlyAuthorizedApplication`
+
+Slashes the provided amount from the stake of every operator in the array.
+Can only be called by an authorized application.
+
+==== `seize(uint256 amount, uint256 rewardMultipier, address notifier, address[] memory operators) external onlyAuthorizedApplication`
+
+Seize the provided amount from the stake of every operator in the array.
+The misbehavior notifier is rewarded with 5% of the total seized amount scaled
+by the reward adjustment parameter. The rest of the tokens are burned. Can only
+be called by an authorized application.
```

Comment:
We might want to make this seize the tokens for the application instead (and convert legacy to T), regardless of what we do on the 5% rule. In that case it should also return the amount transferred.

Reply:
Unsupported in Keep staking contract (untada). See [this](https://github.com/threshold-network/solidity-contracts/pull/15#discussion_r704314021) comment please.